domain/consensus/processes/reachabilitymanager/tree.go

```diff
@@ -317,70 +317,117 @@ func (rt *reachabilityManager) addChild(node, child, reindexRoot *externalapi.Do
 	return rt.stageInterval(child, allocated)
 }
 
-func (rt *reachabilityManager) updateReindexRoot(newTreeNode *externalapi.DomainHash) error {
+func (rt *reachabilityManager) updateReindexRoot(selectedTip *externalapi.DomainHash) error {
 
-	nextReindexRoot, err := rt.reindexRoot()
+	currentReindexRoot, err := rt.reindexRoot()
 	if err != nil {
 		return err
 	}
 
+	// First, find the new root
+	reindexRootAncestor, newReindexRoot, err := rt.findNextReindexRoot(currentReindexRoot, selectedTip)
+	if err != nil {
+		return err
+	}
+
+	// No update to root, return
+	if currentReindexRoot.Equal(newReindexRoot) {
+		return nil
+	}
+
+	rc := newReindexContext(rt)
+
+	// Iterate from reindexRootAncestor towards newReindexRoot
 	for {
-		candidateReindexRoot, found, err := rt.maybeMoveReindexRoot(nextReindexRoot, newTreeNode)
+		chosenChild, err := rt.FindNextAncestor(selectedTip, reindexRootAncestor)
+		if err != nil {
+			return err
+		}
+
+		isFinalReindexRoot := chosenChild.Equal(newReindexRoot)
+
+		// Concentrate interval from current ancestor to it's chosen child
+		err = rc.concentrateInterval(reindexRootAncestor, chosenChild, isFinalReindexRoot)
 		if err != nil {
 			return err
 		}
-		if !found {
+
+		if isFinalReindexRoot {
 			break
 		}
-		nextReindexRoot = candidateReindexRoot
+
+		reindexRootAncestor = chosenChild
 	}
 
-	rt.stageReindexRoot(nextReindexRoot)
+	// Update reindex root data store
+	rt.stageReindexRoot(newReindexRoot)
 	return nil
 }
 
-func (rt *reachabilityManager) maybeMoveReindexRoot(reindexRoot, newTreeNode *externalapi.DomainHash) (
-	newReindexRoot *externalapi.DomainHash, found bool, err error) {
+// findNextReindexRoot finds the new reindex root based on the current one and the new selected tip.
+// The function also returns an ancestor of current and new reindex roots (possibly current root itself).
```

Comment:
Can you write "The common ancestor between the old and new reindex root" instead of "an ancestor"? It'll be clearer